---
title: "GeneOverlap for 2 Wk Vanc RNA Seq Data"
output:
  html_document:
    df_print: paged
  word_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library("GeneOverlap")
library("gdata")
library("VennDiagram")
library("gplots")
library("ggplot2")

library("grid")
library("gridExtra")
library("lattice")


knitr::opts_knit$set(root.dir = "~/code/RnaSeq/ballgown/GeneOverlap")
```



# Loading and inspecting gene lists

Loading 4 gene lists:

* **sham_up** - Genes upregulated in affecteds without treatment when compared to unaffecteds without treatment
```{r}
sham_up <- read.table("../sham/sham_up_gene_results.tsv", sep = "\t", header=TRUE)
head(sham_up)

```

* **sham_down** - Genes downregulated in affecteds without treatment when compared to unaffecteds without treatment
```{r}
sham_down <- read.table("../sham/sham_down_gene_results.tsv", sep = "\t", header=TRUE)
head(sham_down)
```

* **aff_up** - Genes upregulated in affecteds on vancomycin when compared to affecteds without treatment
```{r}
aff_up <- read.table("../aff/aff_up_gene_results.tsv", sep = "\t", header=TRUE)
head(aff_up)
```

* **aff_down** - Genes downregulated in affecteds on vancomycin when compared to affecteds without treatment

```{r geneLists}
aff_down <- read.table("../aff/aff_down_gene_results.tsv", sep = "\t", header=TRUE)
head(aff_down)
```

# Venn Diagrams

Comparing number of overlapping genes between **sham_up** and **aff_down** (i.e. genes upregulated in disease and downregulated with vancomycin)

```{r venn1, echo = FALSE, fig.height=4, fig.width=4, fig.align='center'}
geneLists1 <- list(sham_up$gene_name, aff_down$gene_name)
names(geneLists1) = c("Sham_up", "Aff_down")
VENN.LIST1 <- geneLists1
grid.newpage()
vennplot1 <- venn.diagram(VENN.LIST1, filename = NULL, fill=c("darkmagenta", "darkblue"),
                          alpha=c(0.5,0.5), cex = 2, cat.fontface=4, 
                          category.names=names(geneLists1), cat.cex = 1.5,
                          rotation.degree = 180,
                          main="Sham_up vs Aff_down", main.cex = 2)
```

```{r venn1_plot, echo=FALSE, fig.height=4, fig.width=4, fig.align='center'}
grid::grid.draw(vennplot1)
```


Comparing number of overlapping genes between **sham_down** and **aff_up** (i.e. genes downregulated in disease and upregulated with vancomycin)
```{r venn2, echo = FALSE, fig.height=4, fig.width=4, fig.align='center'}
geneLists2 <- list(sham_down$gene_name, aff_up$gene_name)
names(geneLists2) = c("Sham_down", "Aff_up")
VENN.LIST2 <- geneLists2
grid.newpage()
vennplot2 <- draw.pairwise.venn(length(sham_down$gene_name), length(aff_up$gene_name), length(intersect(sham_down$gene_name, aff_up$gene_name)), 
                                c("Sham_down", "Aff_up"),
                                rotation.degree = 180,
                                fill=c("darkmagenta", "darkblue"),
                                alpha=c(0.5,0.5), 
                                cex = 2, cat.fontface=4, cat.cex = 1.5,
                                main="Sham_down vs Aff_up", main.cex = 2)
```


```{r venn2_plot, echo = FALSE, fig.height=4, fig.width=4, fig.align='center'}
grid::grid.draw(vennplot2)
```

# Simulation

The following simulation calculates the size of overlap between two randomly generate lists from a base of **20,210**, the number of protein encoding genes in the mouse genome (see below). 

A list of **5000** 'genes' (in this case random integers) will randomly be picked from an integer list of 1 to 20,210.

**1000** trials will be run.

```{r}
gs_mouse = 20210
samplesize = 5000
trials = 1000

IntersectionCounts <- sapply(1:trials,
                 function(IntersectCount)
                   c(IntersectCount,length(intersect(sample(gs_mouse, samplesize), sample(gs_mouse, samplesize)))))

# format IntersectionCounts
IntersectionCounts <- as.data.frame(t(IntersectionCounts))
colnames(IntersectionCounts) <- c("Trial","Intersections")
head(IntersectionCounts)
```

The mean intersection size is **`r mean(IntersectionCounts[,"Intersections"])`**. 

```{r venn_simulation, include=FALSE}

# randomList <- replicate(15, sample(gs_mouse, samplesize))

venns <- lapply(1:10, function(i)
  grobTree(draw.pairwise.venn(5000, 5000, IntersectionCounts[i, "Intersections"],
           fill=c("darkmagenta", "darkblue"),
           alpha=c(0.5,0.5),
           cex = 0.5,
           main= paste("Venn", i), main.cex = 2)))

# gs <- lapply(1:9, function(ii) 
#   grobTree(rectGrob(gp=gpar(fill=ii, alpha=0.5)), textGrob(ii)))
# grid.arrange(grobs=gs, ncol=4, 
#                top="top label", bottom="bottom\nlabel", 
#                left="left label", right="right label")
# grid.rect(gp=gpar(fill=NA))
# see: https://cran.r-project.org/web/packages/gridExtra/vignettes/arrangeGrob.html

```


```{r, echo=FALSE, fig.height=2.1, fig.width=5, fig.align='center'}

title=textGrob("Example Random Overlaps", gp=gpar(fontface="bold", fontfamily="Times"))

# tiff(filename = "Venn_simulation.tiff", width = 10, height = 4.3, units = "in", res = 300)

grid.arrange(grobs=venns, ncol=5,
             top=title)
# dev.off()
```

```{r, echo=FALSE, fig.height= 2.6, fig.width = 5, fig.align='center'}
grid.newpage()
vennplot12 <- venn.diagram(VENN.LIST1, filename = NULL, fill=c("darkmagenta", "darkblue"),
                          alpha=c(0.5,0.5), cex = 1, cat.fontface=4, 
                          category.names=names(geneLists1), cat.cex = .5,
                          main="Sham_up vs Aff_down", main.cex = 1)
vennplot22 <- venn.diagram(VENN.LIST2, filename = NULL, fill=c("darkmagenta", "darkblue"),
                          alpha=c(0.5,0.5), cex = 1, cat.fontface=4, 
                          category.names=names(geneLists2), cat.cex = .5,
                          main="Sham_down vs Aff_up", main.cex = 1)
title <- textGrob("For Comparison", gp=gpar(fontfamily="Times",fontface="bold"))
grid.arrange(grobTree(vennplot12), grobTree(vennplot22), ncol=2,
             top=title)
```


# GeneOverlap

Used GeneOverlap package to run statistical testing on whether gene overlaps between group is different than random sampling.

See [link](https://www.bioconductor.org/packages/devel/bioc/vignettes/GeneOverlap/inst/doc/GeneOverlap.pdf) for info on usage of GeneOverlap package used here.

See:
https://stats.stackexchange.com/questions/10328/using-rs-phyper-to-get-the-probability-of-list-overlap
for details on using the hypergeometric distribution to figure out probability of a certain number of overlapping genes

better explanation:
https://rdrr.io/bioc/GeneOverlap/man/GeneOverlap.html

Testing includes **[Fisher's Exact Test](https://en.wikipedia.org/wiki/Fisher%27s_exact_test)** that uses a **hypergeometric distribution** to estimate whether number of overlapping genes is significantly different from random sampling from a fixed genome size.

## Size of Mouse Genome (i.e. sampling pool)

First, we need to establish the number of genes in the mouse genome. 

In the following tests, I will use **20,210** as the number of protein-encoding genes in the mouse genome. I will use this number as it is the smallest and thus most conservative estimation.
[Church et al., 2009](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2680341/)

Jax estimates that there are **24,404** genes with protein sequence data.
[Jax-MGI](http://www.informatics.jax.org/mgihome/homepages/stats/all_stats.shtml#allstats_markers)


Gencode estimates **21,949** protein coding genes.
[gencode](https://www.gencodegenes.org/mouse/stats.html)

```{r}
gs_mouse <- 20210 # Number of protein encoding genes in mouse genome
```


## Comparison of Sham_up and Aff_down

```{r}
GeneOver_sh_up_af_dn <- newGeneOverlap(sham_up$gene_name, aff_down$gene_name, genome.size = gs_mouse)
GeneOver_sh_up_af_dn <- testGeneOverlap(GeneOver_sh_up_af_dn)
print(GeneOver_sh_up_af_dn)
```

p <<< 0.0, so overlap is significantly different as would be expected by random sampling in hypergeometric distribution (tested by Fisher's exact test).



## Comparison of Sham_down and Aff_up

```{r}
GeneOver_sh_dn_af_up <- newGeneOverlap(sham_down$gene_name, aff_up$gene_name, genome.size = gs_mouse)
GeneOver_sh_dn_af_up <- testGeneOverlap(GeneOver_sh_dn_af_up)
print(GeneOver_sh_dn_af_up)
```

p <<< 0.0, so overlap is significantly different as would be expected by random sampling in hypergeometric distribution (tested by Fisher's exact test).


# Comparison of overlap genes with mito genes

Used MitoCarta2.0 to determine list of mito genes. (see [link](https://www.broadinstitute.org/scientific-community/science/programs/metabolic-disease-program/publications/mitocarta/mitocarta-in-0) to download data set used)

```{r, echo = FALSE}
sh_up_af_dn <- intersect(sham_up$gene_name, aff_down$gene_name)
sh_dn_af_up <- intersect(sham_down$gene_name, aff_up$gene_name)

all_DEG <- c(sh_up_af_dn, sh_dn_af_up)
# all_DEG <- as.list(all_DEG)

mito_genes <- read.table("/Users/iskoec/code/TestCode/mito_gene_list.txt", sep = "\n")

mito_genes <- mito_genes$V1

DEG_mitogenes <- intersect(mito_genes, all_DEG)

```

```{r, echo= FALSE}
geneListsMito <- list(all_DEG, mito_genes)
names(geneLists1) = c("all_DEG", "Mito_Genes")
VENN.LISTMITO <- geneListsMito
grid.newpage()
vennplotMito <- venn.diagram(VENN.LISTMITO, filename = NULL, fill=c("darkmagenta", "darkblue"),
                          alpha=c(0.5,0.5), cex = 2, cat.fontface=4, 
                          category.names=names(geneLists1), cat.cex = 1.5,
                          main="all_DEG vs Mito genes", main.cex = 2)
```

```{r vennmito_plot, echo=FALSE, fig.height=4, fig.width=4, fig.align='center'}
grid::grid.draw(vennplotMito)
```

The number of mito genes that are observed in vancomycin correction (combined overlaps) is 
**`r length(DEG_mitogenes)`**.



### Is overlap between mito genes and overall genes significant?

```{r}
GeneOver_mito <- newGeneOverlap(all_DEG, mito_genes, genome.size = gs_mouse)
GeneOver_mito <- testGeneOverlap(GeneOver_mito)
print(GeneOver_mito)
```

Well, p <<< 0.001, but jaccard index (intersection/union) is a little low (could be due to differences in size of gene lists).


